Mathlib/Tactic/Backtracking.lean

```diff
@@ -207,4 +207,375 @@ where
+`exceeds` another, implying that the `β` with the excessive size can be discarded as non-minimal.
+We also give a means to `incorporate` elements of `β` together when multiple
+-/
+structure BacktrackMinimizeListConfig (m : Type _ → Type _) [Monad m] [Alternative m]
+    [MonadBacktrack σ m] (α β : Type _) (γ : Type v) where
+  /-- The maximum recursion depth. -/
+  maxDepth : Nat := 6
+  /-- An arbitrary procedure which can be used to modify the list of `α`s
+  before each attempt to apply alternatives.
+  Called as `proc init curr`, where `init` is the initial list of `α`s,
+  and `curr` are the current list of `α`s.
+  Returning `some l` will replace the current `α`s with `l` just before alternatives are generated.
+  Note that this does not consume a step of maximum depth.
+  Returning `none` will proceed to generating alternatives.
+  Failure will cause the whole list to fail and cause backtracking.
+  (Defaults to returning `none`.) -/
+  proc : List α → List α → m (Option (List α)) := fun _ _ => pure none
+  /-- The minimal element of `β` to begin the search with. This should be a left identity for
+  `incorporate`, i.e. `incorporate bot a = a`. -/
+  bot : β
+  /-- A function to calculate the size of a possible best element (`β`), where the notion of size
+  is given by `γ` and `exceeds`. This function should be cheap. -/
+  size : β → γ
+  /-- An operation to add two sizes together. `size` should distribute over
+  `incorporate : β → β → β` via `add : γ → γ → γ`; i.e., we should have
+  `add (size a) (size b) = size (incorporate a b)`. -/
+  add : γ → γ → γ
+  /-- `exceeds (size newBestCandidate) (size oldBest)` should be `true` exactly when
+  `newBestCandidate` is too big to be considered a minimum. This implies that the
+  `newBestCandidate` is not actually viable, and will cause backtracking and/or movement to the
+  next alternative.
+  -/
+  exceeds : γ → γ → Bool
+  /-- `takeBestOfEq oldBest newBestCandidate` is used only when two candidates `newBestCandidate`,
+  `oldBest` are "equal" in the sense that neither exceed the other; that is, both
+  `exceeds (size newBestCandidate) (size oldBest)` and
+  `exceeds (size oldBest) (size newBestCandidate)` are `false`. Note that if `exceeds` behaves like
+  `≥`, `takeBestOfEq` will never be reached. In this case, `takeBestOfEq` should be set to `none`.
+  However, if `exceeds` behaves like `>`, `takeBestOfEq` may be reached. It is always called with
+  the older candidate in the first argument and the newer one in the second. By default, this is
+  `none`, which, if encountered, simply takes the first `best`. -/
+  takeBestOfEq? : Option (σ × β → σ × β → m β) := none
+  /-- `incorporate current localBest` represents the "addition" of two necessary components of a
+  minimal element. This is called to consolidate the `current` minimal elements obtained from the
+  list currently being examined with the `localBest` generated from the next element in the list. -/
+  incorporate : β → β → β
```

Comment:
I have to admit that this whole API for working with sizes feels baroque, and I would prefer to see something much simpler, or failing that some explanation of why something simpler wouldn't suffice. :-)

Reply:
That's an entirely fair sentiment 🙃 The two motivations for doing it this way were

1) process-based: by working abstractly (instead of, say, just using lists or nats), I could be more sure that I was writing correct code, and could be sure of which parts of the data I actually needed.

2) generality: there are a couple of imaginable cases where you want to do the same process, but want to use different notions of `exceeds` and different notions of sizes. For example, we're able to use `size` on an `Option α` (`1` for `some _`, `0` for `none`) to do things more efficiently for `minimizeItems`. That's about it for now, admittedly—but this lets you in principle do things like
* gather *all* of the "best" sequences (instead of just the first via using a `>`-like notion of `exceeds` with `takeBestOfEq`)
* extend `γ` to types generalizing `Nat` for better control, e.g. `WithTop Nat` to consider some elements worse (bigger) than everything else, but better than nothing
* or do things I can't anticipate, like, I don't know, finding out which collection of matrices attains a minimal determinant when multiplied together (`size`  = `det`, `incorporate` = matrix multiplication, `add` = multiplication

That being said, I'm willing to restrict the capabilities, but I don't have a good idea as to what the "right" way to restrict it would be. Open to ideas! :)